File: C-Sharp/01_Basics/08_methods.cs
// ─────────────────────────────────────────
//  METHODS (FUNCTIONS)
// ─────────────────────────────────────────

using System;
using System.Collections.Generic;

class Methods
{
    // ── Basic method ──────────────────────────
    static string Greet(string name)
    {
        return $"Hello, {name}!";
    }

    // ── void — no return value ────────────────
    static void PrintLine(string text)
    {
        Console.WriteLine(text);
    }

    // ── Default parameters ────────────────────
    static string Greet2(string name, string greeting = "Hello")
    {
        return $"{greeting}, {name}!";
    }

    // ── Named arguments ───────────────────────
    static void CreateUser(string name, int age, string city = "Unknown")
    {
        Console.WriteLine($"{name}, {age}, {city}");
    }

    // ── Return multiple values (tuple) ────────
    static (int Min, int Max, double Avg) GetStats(int[] nums)
    {
        return (Array.Min(nums), Array.Max(nums), nums.Average());
    }
    // helper since Array doesn't have Min/Max static — using LINQ:
    static (int min, int max) MinMax(int[] nums)
    {
        int min = nums[0], max = nums[0];
        foreach (var n in nums) { if (n < min) min = n; if (n > max) max = n; }
        return (min, max);
    }

    // ── params — variable number of arguments ─
    static int Sum(params int[] numbers)
    {
        int total = 0;
        foreach (var n in numbers) total += n;
        return total;
    }

    // ── ref — pass by reference ───────────────
    static void Increment(ref int value)
    {
        value++;   // modifies the original variable
    }

    // ── out — return via parameter ────────────
    static bool TryDivide(int a, int b, out double result)
    {
        if (b == 0) { result = 0; return false; }
        result = (double)a / b;
        return true;
    }

    // ── Expression-bodied method (C# 6+) ──────
    static int Square(int x) => x * x;
    static string Repeat(string s, int n) => string.Concat(Enumerable.Repeat(s, n));

    // ── Method overloading ────────────────────
    static int Add(int a, int b) => a + b;
    static double Add(double a, double b) => a + b;
    static string Add(string a, string b) => a + b;

    // ── Recursive method ──────────────────────
    static int Factorial(int n)
    {
        if (n <= 1) return 1;
        return n * Factorial(n - 1);
    }

    static int Fibonacci(int n)
    {
        if (n <= 1) return n;
        return Fibonacci(n - 1) + Fibonacci(n - 2);
    }

    static void Main()
    {
        // Basic call
        Console.WriteLine(Greet("Alice"));          // Hello, Alice!
        PrintLine("I print nothing back");

        // Default parameters
        Console.WriteLine(Greet2("Bob"));            // Hello, Bob!
        Console.WriteLine(Greet2("Bob", "Hi"));      // Hi, Bob!

        // Named arguments — any order
        CreateUser(age: 25, name: "Alice");
        CreateUser("Bob", city: "NYC", age: 30);

        // Multiple return
        var (min, max) = MinMax(new[] { 3, 1, 7, 2, 9 });
        Console.WriteLine($"Min={min}, Max={max}");

        // params
        Console.WriteLine(Sum(1, 2, 3));         // 6
        Console.WriteLine(Sum(1, 2, 3, 4, 5));   // 15

        // ref
        int val = 5;
        Increment(ref val);
        Console.WriteLine(val);   // 6

        // out
        if (TryDivide(10, 3, out double result))
            Console.WriteLine($"10/3 = {result:F2}");

        TryDivide(10, 0, out double r2);
        Console.WriteLine(r2);   // 0 — didn't crash

        // Expression-bodied
        Console.WriteLine(Square(7));         // 49
        Console.WriteLine(Repeat("ab", 3));   // ababab

        // Overloading
        Console.WriteLine(Add(2, 3));           // 5  (int)
        Console.WriteLine(Add(2.5, 3.5));       // 6  (double)
        Console.WriteLine(Add("Hello", " C#")); // Hello C# (string)

        // Recursion
        Console.WriteLine(Factorial(5));   // 120
        Console.WriteLine(Factorial(10));  // 3628800

        for (int i = 0; i < 8; i++)
            Console.Write(Fibonacci(i) + " ");   // 0 1 1 2 3 5 8 13
        Console.WriteLine();

        // ── Local functions ───────────────────────
        int MultiplyBy(int x, int factor) => x * factor;
        Console.WriteLine(MultiplyBy(5, 3));   // 15

        // ── Lambda / Anonymous functions ──────────
        Func<int, int> doubleIt = x => x * 2;
        Func<int, int, int> multiply = (a, b) => a * b;
        Action<string> shout = s => Console.WriteLine(s.ToUpper());

        Console.WriteLine(doubleIt(7));        // 14
        Console.WriteLine(multiply(3, 4));     // 12
        shout("hello");                        // HELLO

        // Lambdas with LINQ
        var nums = new[] { 1, 2, 3, 4, 5, 6 };
        var evens = System.Linq.Enumerable.Where(nums, x => x % 2 == 0);
        foreach (var n in evens) Console.Write(n + " ");   // 2 4 6
        Console.WriteLine();
    }
}

File: Java/01_Basics/08_Methods.java
// ─────────────────────────────────────────
//  METHODS
// ─────────────────────────────────────────

import java.util.*;
import java.util.function.*;

public class Methods {

    // ── Basic method ──────────────────────────
    static String greet(String name) {
        return "Hello, " + name + "!";
    }

    // ── void — no return value ────────────────
    static void printLine(String text) {
        System.out.println(text);
    }

    // ── Return multiple values — use array, list, or record ─
    static int[] minMax(int[] nums) {
        int min = nums[0], max = nums[0];
        for (int n : nums) {
            if (n < min) min = n;
            if (n > max) max = n;
        }
        return new int[]{min, max};
    }

    // ── Varargs — variable number of arguments ─
    static int sum(int... numbers) {
        int total = 0;
        for (int n : numbers) total += n;
        return total;
    }

    // Can combine with regular params (varargs must be last)
    static String join(String sep, String... words) {
        return String.join(sep, words);
    }

    // ── Method overloading ────────────────────
    static int add(int a, int b)          { return a + b; }
    static double add(double a, double b) { return a + b; }
    static String add(String a, String b) { return a + b; }

    // ── Recursive methods ─────────────────────
    static long factorial(int n) {
        if (n <= 1) return 1;
        return n * factorial(n - 1);
    }

    static int fibonacci(int n) {
        if (n <= 1) return n;
        return fibonacci(n - 1) + fibonacci(n - 2);
    }

    // ── Generic method ────────────────────────
    static <T> void printArray(T[] arr) {
        System.out.print("[");
        for (int i = 0; i < arr.length; i++) {
            System.out.print(arr[i]);
            if (i < arr.length - 1) System.out.print(", ");
        }
        System.out.println("]");
    }

    static <T extends Comparable<T>> T findMax(T[] arr) {
        T max = arr[0];
        for (T item : arr)
            if (item.compareTo(max) > 0) max = item;
        return max;
    }

    public static void main(String[] args) {

        // Basic calls
        System.out.println(greet("Alice"));    // Hello, Alice!
        printLine("printed");

        // Multiple return
        int[] result = minMax(new int[]{3, 1, 7, 2, 9});
        System.out.println("min=" + result[0] + ", max=" + result[1]);

        // Varargs
        System.out.println(sum(1, 2, 3));          // 6
        System.out.println(sum(1, 2, 3, 4, 5));    // 15
        System.out.println(join("-", "a", "b", "c")); // a-b-c

        // Overloading
        System.out.println(add(2, 3));             // 5 (int)
        System.out.println(add(2.5, 3.5));         // 6.0 (double)
        System.out.println(add("Hello", " Java")); // Hello Java

        // Recursion
        System.out.println(factorial(5));  // 120
        System.out.println(factorial(10)); // 3628800

        for (int i = 0; i < 8; i++)
            System.out.print(fibonacci(i) + " ");  // 0 1 1 2 3 5 8 13
        System.out.println();

        // Generics
        Integer[] ints = {3, 1, 7, 2};
        String[] strs = {"banana", "apple", "cherry"};
        printArray(ints);
        printArray(strs);
        System.out.println(findMax(ints));   // 7
        System.out.println(findMax(strs));   // cherry

        // ── Functional interfaces & lambdas ───────
        // Function<T, R> — takes T, returns R
        Function<String, Integer> strLen = s -> s.length();
        System.out.println(strLen.apply("hello"));   // 5

        // BiFunction<T, U, R> — takes T and U, returns R
        BiFunction<Integer, Integer, Integer> multiply = (a, b) -> a * b;
        System.out.println(multiply.apply(3, 4));    // 12

        // Consumer<T> — takes T, returns void
        Consumer<String> shout = s -> System.out.println(s.toUpperCase());
        shout.accept("hello");   // HELLO

        // Supplier<T> — takes nothing, returns T
        Supplier<String> greeting = () -> "Hello, World!";
        System.out.println(greeting.get());

        // Predicate<T> — takes T, returns boolean
        Predicate<Integer> isEven = n -> n % 2 == 0;
        System.out.println(isEven.test(4));   // true
        System.out.println(isEven.test(7));   // false

        // Predicate chaining
        Predicate<Integer> isPositive = n -> n > 0;
        Predicate<Integer> isPositiveEven = isEven.and(isPositive);
        System.out.println(isPositiveEven.test(4));   // true
        System.out.println(isPositiveEven.test(-4));  // false

        // ── Method references ──────────────────────
        Function<String, String> upper = String::toUpperCase;  // instance method
        System.out.println(upper.apply("hello"));   // HELLO

        Consumer<String> printer = System.out::println;        // static method ref
        printer.accept("Printed via method ref");

        List<String> words = List.of("banana", "apple", "cherry");
        words.stream()
             .map(String::toUpperCase)
             .forEach(System.out::println);
    }
}
